torchvision/prototype/datasets/_builtin/food101.py

```diff
@@ -0,0 +1,86 @@
+import functools
+from pathlib import Path
+from typing import Any, Tuple, List, Dict, cast, Optional, BinaryIO
+
+from torchdata.datapipes.iter import IterDataPipe, Filter, Mapper, CSVParser, JsonParser, Demultiplexer
+from torchvision.prototype.datasets.utils import Dataset, DatasetInfo, DatasetConfig, HttpResource, OnlineResource
+from torchvision.prototype.datasets.utils._internal import (
+    hint_shuffling,
+    hint_sharding,
+    path_comparator,
+    getitem,
+    INFINITE_BUFFER_SIZE,
+)
+from torchvision.prototype.features import Label, EncodedImage
+
+
+class Food101(Dataset):
+    def _make_info(self) -> DatasetInfo:
+        return DatasetInfo(
+            "food101",
+            homepage="https://data.vision.ee.ethz.ch/cvl/datasets_extra/food-101",
+            # Only train and test splits, no validation.
+            valid_options=dict(split=("train", "test")),
+        )
+
+    def resources(self, config: DatasetConfig) -> List[OnlineResource]:
+        return [
+            HttpResource(
+                url="http://data.vision.ee.ethz.ch/cvl/food-101.tar.gz",
+                sha256="d97d15e438b7f4498f96086a4f7e2fa42a32f2712e87d3295441b2b6314053a4",
+                decompress=True,
+            )
+        ]
+
+    def _classify_archive(self, data: Tuple[str, Any]) -> Optional[int]:
+        path = Path(data[0])
+        if path.parents[1].name == "images":
+            return 0
+        elif path.parents[0].name == "meta":
+            return 1
+        else:
+            return None
+
+    def _prepare_sample(self, data: Tuple[str, BinaryIO]) -> Dict[str, Any]:
+        image_path, image_buffer = data
+        category = Path(image_path).parent.name
+        return dict(
+            image=EncodedImage.from_file(image_buffer),
+            label=Label.from_category(category, categories=self.categories),
+        )
```

Comment:
Small changes for the suggestion above.

```suggestion
    def _prepare_sample(self, data: Tuple[str, Tuple[str, BinaryIO]]) -> Dict[str, Any]:
        id, (path, buffer) = data
        return dict(
            label=Label.from_category(id.split("/", 1)[0], categories=self.categories),
            path=path,
            image=EncodedImage.from_file(buffer),
        )
```

Reply:
Questions: is returning the `path` something recommended? Also, shouldn't the `id` variable be given another name (since id is an internal Python variable) or is that fine here given the context? Thanks.

Reply:
> is returning the `path` something recommended?

Yes. We had a couple of user requests asking us to give them the ability to "perform surgery" on failing examples. For example, if your pipeline fails on a specific example, you might want to look at the file to see what is going on. Without the path, this is impossible.

> Also, shouldn't the `id` variable be given another name (since id is an internal Python variable) or is that fine here given the context?

We can, but this is nothing major. The overwrite is constrained to 5 lines and TBH I've never used `id()` anyway. I'm open to suggestions though.

Reply:
I see. Makes sense to suggest having access to the path variable for debug. id should be fine indeed, I have almost never used `id()` as well. :smile_cat: